File: app/src/main/java/com/angogasapps/myfamily/ui/screens/buy_list/adapters/BuyListAdapter.java
package com.angogasapps.myfamily.ui.screens.buy_list.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.angogasapps.myfamily.R;
import com.angogasapps.myfamily.databinding.BuyListHolderBinding;
import com.angogasapps.myfamily.models.BuyList;
import com.angogasapps.myfamily.models.BuyListEvent;
import com.angogasapps.myfamily.ui.screens.buy_list.BuyListActivity;
import com.angogasapps.myfamily.ui.screens.buy_list.BuyListManager;
import com.angogasapps.myfamily.ui.screens.buy_list.dialogs.ChangeOrDeleteBuyListDialog;

import java.util.ArrayList;

public class BuyListAdapter extends RecyclerView.Adapter<BuyListAdapter.BuyListHolder> {
    private Context context;
    private LayoutInflater inflater;
    private ArrayList<BuyList> buyListsArray;

    {
        buyListsArray = BuyListManager.getInstance().getBuyLists();
    }

    public BuyListAdapter(Context context){
        this.context = context;
        this.inflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public BuyListHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new BuyListHolder(inflater.inflate(R.layout.buy_list_holder, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull BuyListHolder holder, int position) {
        holder.setTextName(buyListsArray.get(position).getName());
        holder.binding.getRoot().setOnLongClickListener(v -> {
            (new ChangeOrDeleteBuyListDialog(context, buyListsArray.get(position))).show();
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return buyListsArray.size();
    }


    public void updateEnd(){
        notifyItemInserted(buyListsArray.size() - 1);
    }

    public void update(BuyListEvent event) {
        if (event.getEvent().equals(BuyListEvent.EBuyListEvents.buyListRemoved)){
            this.notifyItemRemoved(event.getIndex());
        }
        if (event.getEvent().equals(BuyListEvent.EBuyListEvents.buyListAdded)){
            this.notifyItemChanged(event.getIndex());
        }
        if (event.getEvent().equals(BuyListEvent.EBuyListEvents.buyListChanged)){
            this.notifyItemChanged(event.getIndex());
        }
    }

    public class BuyListHolder extends RecyclerView.ViewHolder {
        private BuyListHolderBinding binding;
        public BuyListHolder(@NonNull View itemView) {
            super(itemView);
            binding = BuyListHolderBinding.bind(itemView);

            binding.getRoot().setOnClickListener(v -> {
                BuyListActivity.getIGoToBuyListFragment().go(BuyListAdapter.this.buyListsArray.get(getLayoutPosition()));
            });
        }

        public void setTextName(String textName){
            binding.textName.setText(textName);
        }
    }
}

File: app/src/main/java/com/angogasapps/myfamily/ui/screens/buy_list/adapters/BuyListAdapter.kt
package com.angogasapps.myfamily.ui.screens.buy_list.adapters

import android.content.Context
import com.angogasapps.myfamily.ui.screens.buy_list.BuyListActivity.Companion.iGoToBuyListFragment
import androidx.recyclerview.widget.RecyclerView
import com.angogasapps.myfamily.ui.screens.buy_list.adapters.BuyListAdapter.BuyListHolder
import android.view.LayoutInflater
import android.view.View
import com.angogasapps.myfamily.models.buy_list.BuyList
import android.view.ViewGroup
import com.angogasapps.myfamily.R
import com.angogasapps.myfamily.databinding.BuyListHolderBinding
import com.angogasapps.myfamily.ui.screens.buy_list.dialogs.ChangeOrDeleteBuyListDialog
import com.angogasapps.myfamily.models.buy_list.BuyListEvent
import java.util.ArrayList

class BuyListAdapter(private val context: Context, private val buyLists: ArrayList<BuyList>) :
    RecyclerView.Adapter<BuyListHolder>() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private var buyListsArray: ArrayList<BuyList> = buyLists
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BuyListHolder {
        return BuyListHolder(inflater.inflate(R.layout.buy_list_holder, parent, false))
    }

    override fun onBindViewHolder(holder: BuyListHolder, position: Int) {
        synchronized(this) {
            holder.setTextName(buyListsArray[position].title)
            holder.binding.root.setOnLongClickListener {
                ChangeOrDeleteBuyListDialog(context, buyListsArray[holder.position]).show()
                true
            }
        }
    }

    override fun getItemCount(): Int {
        return buyListsArray.size
    }

    fun updateEnd() {
        notifyItemInserted(buyListsArray.size - 1)
    }

    fun update(event: BuyListEvent) {
        when (event.event) {
            BuyListEvent.EBuyListEvents.buyListRemoved -> {
                notifyItemRemoved(event.index)
            }
            BuyListEvent.EBuyListEvents.buyListAdded -> {
                this.notifyItemChanged(event.index)
            }
            BuyListEvent.EBuyListEvents.buyListChanged -> {
                this.notifyItemChanged(event.index)
            }
        }
    }

    inner class BuyListHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val binding: BuyListHolderBinding = BuyListHolderBinding.bind(itemView)

        fun setTextName(textName: String) {
            binding.textName.text = textName
        }

        init {
            binding.root.setOnClickListener {
                iGoToBuyListFragment.invoke(buyListsArray[layoutPosition])
            }
        }
    }
}